channeld/channeld.c

```diff
@@ -3896,6 +3915,9 @@ static void init_channel(struct peer *peer)
 		master_badmsg(WIRE_CHANNELD_INIT, msg);
 	}
 
+	peer->final_index = tal_dup(peer, u32, &final_index);
```

Comment:
these fields seem to be non-optional in `channeld_init`, but they are pointers in `struct peer`, so I'm confused.

Reply:
They need to be optional in `struct peer` because they are used by `penalty_tx_create` which needs optional.

Indeed, they are not optional in `channeld_init`, the default is always supplied at the beginning.

But after that, a `channeld_send_shutdown` may be handled, which needs to be able to set the `final_index` to NULL if a late (non-upfront) shutdown script was supplied and it's not in the wallet.

Reply:
Got it